pkg/storage/ingest/pusher.go

```diff
@@ -103,15 +103,57 @@ func newPusherConsumer(pusher Pusher, kafkaCfg KafkaConfig, metrics *pusherConsu
 	}
 }
 
-func (c pusherConsumer) consume(ctx context.Context, records []record) error {
-	ctx, cancel := context.WithCancelCause(ctx)
-	defer cancel(cancellation.NewErrorf("done consuming records"))
+// Consume implements the recordConsumer interface.
+// It'll use a separate goroutine to unmarshal the next record while we push the current record to storage.
+func (c pusherConsumer) Consume(ctx context.Context, records []record) error {
+	recordsChannel := make(chan parsedRecord)
 
-	recC := make(chan parsedRecord)
+	// Create a cancellable context to let the unmarshalling goroutine know when to stop.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	// Speed up consumption by unmarhsalling the next request while the previous one is being pushed.
-	go c.unmarshalRequests(ctx, records, recC)
-	return c.pushRequests(recC)
+	// Now, unmarshal the records into the channel.
+	go func(unmarshalCtx context.Context, records []record) {
+		defer close(recordsChannel)
+
+		for index, r := range records {
+			select {
+			case <-unmarshalCtx.Done():
+				// No more processing is needed, so we need to abort.
+				return
+			default:
+				parsed := parsedRecord{
+					ctx:          r.ctx,
+					tenantID:     r.tenantID,
+					WriteRequest: &mimirpb.WriteRequest{},
+					index:        index,
+				}
+
+				// We don't free the WriteRequest slices because they are being freed by a level below.
+				err := parsed.WriteRequest.Unmarshal(r.content)
+				if err != nil {
+					parsed.err = fmt.Errorf("parsing ingest consumer write request: %w", err)
+				}
+				recordsChannel <- parsed
+			}
+		}
+	}(ctx, records)
```

Comment:
we pass `recordsChannel` by closure and `ctx` and `records` by parameter. Should we do it just one way?